Marc/modbus polling

CanuckMarc opened the pull request:
@RaiBnod @NubeDev 
This is a tested implementation of modbus with the new polling queue. 

You guys should review a couple of parts of the code changes:
- Changes to the flow of `PointWrite` to be via the plugins.  Changes in `api`.
- Changes to `dbhandler` functions.
- Changes to `database` functions.

Below are notes regarding the completed tasks and still to complete tasks.  This code is functional and can be merged for further testing. 

- [x] - Poll queue adds all points to PriorityPollQueue on modbus plugin enable
- [x] - QueueUnloader automatically posts the next polling point for each network queue
- [x] - Polling complete callback works, and re-adds points to the polling queue based on writeMode (with timer)
- [x] - Poll queue doesn't necessarily keep the points in priority order, but it will always draw the highest priority from the queue first.  
- [x] - Added another slice to keep track on standby points- eventbus triggers for add remove and update network/device/point
- [x] - add in a point property that indicates that the point PresentValue should only be updated from the polling (not from the priority array).
- [x] - add UI selections to model.go in modbus plugin. 
- [x] - test writeMode types.
- [x] - add ping check for modbus port availability, otherwise the polling races with failed calls. 
- [x] - implement a system to promote points to be polled ASAP (on write, or when manually triggered). 
- [x] - test disable/enable of network/device/point


- [ ]  time.Duration types in settings show up in ns. 
- [ ]  periodically check the polling queue and standby points to check that none have been lost.  
- [ ]  implement checks/monitoring statistics to show if points are not being polled due to short cycling of higher priority points. - 
- [ ]  check that point faults are set and cleared correctly.
- [ ]  test changing of polling priority, poll rate (at device and point level)
- [ ] saving polling queues between enable/disable
- [ ] optimize modbus polling to query multiple points as one request (complicated and might depend on the device)
- [ ] modbus polling currently loops through networks, but the modbus client may be able to support a seperate go routine for each network. This would increase the polling rate significantly.
- [ ]  risk of short cycling error points with higher priority than lower priority points. 
- [ ] Creating a client on each loop is bad practice, we should fix this up for efficiency.
- [ ] Should write values should be given a higher priority in the poll queue.  I think probably.  High Priority Writes -> High Priority Reads -> Normal Priority Writes -> Normal Priority Reads -> etc

NubeDev commented:
these are kinda unsable for defults 
probs best to show the user that they are `ms` if they are aswell

![image](https://user-images.githubusercontent.com/18637436/166091318-61df5442-f50f-469b-95df-59e0e4295cbd.png)

CanuckMarc commented:
@NubeDev I got kinda carried away with find and replace on `(i *Instance) --> (inst *Instance)` and ended up changing it throughout FF.  I figured it would be good if it was standardized anyways, but let me know if you want me to undo that commit and just fix the ones in my modbus polling relevant code.

CanuckMarc commented:
@NubeDev regarding the polling times, they are currently set in `ns` because that is what GO `time.Duration` types resolve to.  I debated manually changing it to just use seconds and `int` types.  But I think we should probably figure out a way to handle our time input UIs so that it works with `time.Duration` types (as they are easier to understand when work with them in code).  There will be MANY requirements for time based settings, so we should get a UI solution that works well. 

For now, I figured we could just deal with `ns`.  But if the UI improvement would be too delayed, then i'll just switch it up.  Just requires changing some types on the Point/Device/Network models.  Let me know.

NubeDev commented:
@RaiBnod 
Can we look at merging this

there is a conflict ill get marc to fix that
@CanuckMarc pls fix

RaiBnod commented on database/point_plugin.go:
```diff
@@ -56,12 +56,34 @@ func (d *GormDatabase) UpdatePointPlugin(uuid string, body *model.Point) (point
+		return nil, err
+	}
+	pluginName := network.PluginPath
+	if pluginName == "system" {
+		point, err = d.PointWrite(body.UUID, body, false)
+		if err != nil {
+			return nil, err
+		}
+		return
+	}
+
+	cli := client.NewLocalClient()
+	point, err = cli.WritePointPlugin(uuid, body, pluginName)
+	if err != nil {
+		return nil, err
+	}
```
For plugins that are other than "system" and "modbus", they won't find the API to write the value. Coz, other plugins don't have that `PATCH` write API on its plugin.

So, maybe for now we just need to include that `cli.WritePointPlugin(...)` for "modbus".

RaiBnod commented on plugin/nube/protocals/modbus/polling.go:
```diff
@@ -36,116 +38,236 @@ func delays(networkType string) (deviceDelay, pointDelay time.Duration) {
+			var mbClient smod.ModbusClient
+			//var dCheck devCheck
+			//dCheck.devUUID = dev.UUID
+			mbClient, err = inst.setClient(net, dev, true)
```
Creating a client on each loop will be bad. If we are planning to improve this one later, it's all good for now.

CanuckMarc replied:
Noted.  I'll add it to my checklist of items to  be addressed still.

RaiBnod commented on database/point_plugin.go:
```diff
@@ -56,12 +56,34 @@ func (d *GormDatabase) UpdatePointPlugin(uuid string, body *model.Point) (point
 	return
 }
 
+func (d *GormDatabase) WritePointPlugin(uuid string, body *model.Point) (point *model.Point, err error) {
+	network, err := d.GetNetworkByPointUUID(uuid, api.Args{})
+	if err != nil || network == nil {
+		return nil, err
+	}
+	pluginName := network.PluginPath
+	if pluginName == "system" {
+		point, err = d.PointWrite(body.UUID, body, false)
+		if err != nil {
+			return nil, err
+		}
+		return
+	}
+
+	cli := client.NewLocalClient()
+	point, err = cli.WritePointPlugin(uuid, body, pluginName)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
```
For plugins that are other than "system" and "modbus", they won't find the API to write the value. Coz, other plugins don't have that `PATCH` write API on its plugin.

So, maybe for now we just need to include that `cli.WritePointPlugin(...)` for "modbus". Or have that write API implemented on other plugins as well.

RaiBnod commented on plugin/nube/protocals/modbus/poll-queue/poll-queue.go:
```diff
@@ -0,0 +1,313 @@
+	nq.RemoveDeviceFromActiveDevicesList(deviceUUID)
+	return true
+}
+func (nq *NetworkPriorityPollQueue) UpdatePollingPointByPointUUID(pointUUID string, newPriority model.PollPriority) bool {
+	nq.PriorityQueue.UpdatePollingPointByPointUUID(pointUUID, newPriority)
+	nq.StandbyPollingPoints.UpdatePollingPointByPointUUID(pointUUID, newPriority)
+	return true
+}
```
Seems like this never gets called for fixing the priority-wise unloading. Is this priority-wise unloading is yet to be implemented?

RaiBnod commented on database/point_plugin.go:
```diff
@@ -56,12 +56,34 @@ func (d *GormDatabase) UpdatePointPlugin(uuid string, body *model.Point) (point
 	return
 }
 
+func (d *GormDatabase) WritePointPlugin(uuid string, body *model.Point) (point *model.Point, err error) {
```
For plugins that are other than "system" and "modbus", they won't find the API to write the value. Coz, other plugins don't have that `PATCH` write API on its plugin.

So, maybe for now we just need to include that `cli.WritePointPlugin(...)` for "modbus". Or have that write API implemented on other plugins as well.

CanuckMarc replied:
Good catch.  I'll add default `writePoint` handlers to the other plugins, or add an exception for `modbus`.

RaiBnod reviewed (changes requested):
I have just gone through the code surfacely and added some of the concerns.

CanuckMarc commented:
@RaiBnod @NubeDev 

I've updated all the protocol plugins so that `write point` call flows through the plugin before making the DB call.  

I also fixed the merge errors.

RaiBnod commented on plugin/nube/protocals/modbus/poll-queue/poll-manager.go:
```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+func NewPollManager(dbHandler *dbhandler.Handler, FFNetworkUUID, FFPluginUUID string) *NetworkPollManager {
```
The variable on go never starts for caps in go, so you better to refactor this: as ffNetworkUUID, ffPluginUUID. I've seen such variables in other places as well.

CanuckMarc replied:
ok, i'll do that now.  Aidan found one of those the other day too.

CanuckMarc replied:
done

RaiBnod commented on plugin/nube/protocals/modbus/poll-queue/poll-manager.go:
```diff
@@ -0,0 +1,224 @@
+
+// Polling Manager Summary:
+//  - Diagram Summary: https://docs.google.com/drawings/d/1priwsaQ6EryRBx1kLQd91REJvHzFyxz7cOHYYXyBNFE/edit?usp=sharing
+//  - The PollManager Inserts, Removes, and Updates PollingPoints from the PriorityPollQueue based on the settings of the point.  It considers the Network/Device/Point Enables, Point Write-Modes, etc.
```
Please break such comments into multi-lines, the standard is:
- It needs to have codes or comments max till 120 characters in a line

This is because on a small screen. It will be really hard for the readability.

CanuckMarc replied:
removed. and noted.
